integration/service/create_test.go

```diff
@@ -309,6 +312,144 @@ func TestCreateServiceConfigFileMode(t *testing.T) {
+func TestCreateServiceSysctls(t *testing.T) {
+	defer setupTest(t)()
+	d := swarm.NewSwarm(t, testEnv)
+	defer d.Stop(t)
+	client := d.NewClientT(t)
+	defer client.Close()
+
+	ctx := context.Background()
+
+	// run thie block twice, so that no matter what the default value of
+	// net.ipv4.ip_nonlocal_bind is, we can verify that setting the sysctl
+	// options works
+	for _, expected := range []string{"0", "1"} {
+
+		// store the map we're going to be using everywhere.
+		sysctlOpts := map[string]string{"net.ipv4.ip_nonlocal_bind": expected}
+
+		// Create the service with the sysctl options
+		var instances uint64 = 1
+		serviceName := "TestService_" + t.Name() + "_" + expected
+		serviceID := swarm.CreateService(t, d,
+			swarm.ServiceWithName(serviceName),
+			swarm.ServiceWithCommand([]string{
+				"/bin/sh", "-c",
+				"cat /proc/sys/net/ipv4/ip_nonlocal_bind && sleep 10",
+			}),
+			swarm.ServiceWithReplicas(instances),
+			swarm.ServiceWithSysctls(sysctlOpts),
+		)
+
+		// wait for the service to converge to 1 running task as expected
+		poll.WaitOn(t, serviceRunningTasksCount(client, serviceID, instances))
+
+		// get all of the tasks of the service, so we can get the container
+		// logs
+		filter := filters.NewArgs()
+		filter.Add("service", serviceID)
+		tasks, err := client.TaskList(ctx, types.TaskListOptions{
+			Filters: filter,
+		})
+		assert.NilError(t, err)
+		assert.Check(t, is.Equal(len(tasks), 1))
+
+		// just because the container started, doesn't mean it's actually run
+		// the command yet. if we try to read the logs instantly, we might be
+		// too fast and read before the output of the cat is available. to
+		// avoid that case, we're using poll.WaitOn with a closure to wait
+		// until logs actually respond
+		poll.WaitOn(t, func(t poll.LogT) poll.Result {
+			body, err := client.ContainerLogs(
+				ctx, tasks[0].Status.ContainerStatus.ContainerID,
+				types.ContainerLogsOptions{ShowStdout: true},
+			)
+			if err != nil {
+				return poll.Error(err)
+			}
+			defer body.Close()
+			// when we get logs, they come multiplexed into a single stream, for
+			// both stdout and stderr. in order to properly check the return value,
+			// we need to demultiplex these streams using stdcopy. The first arg is
+			// a bytes buffer we're creating to get everything out. The second arg
+			// is for stderr, which we don't care about and are throwing away. The
+			// third arg is the response we just got that we're reading from.
+			buf := &bytes.Buffer{}
+			if _, err := stdcopy.StdCopy(buf, ioutil.Discard, body); err != nil {
+				return poll.Error(err)
+			}
+
+			// get the string representation of the buffer, and trim off any extra
+			// whitespace that may have gotten in.
+			val := strings.TrimSpace(buf.String())
+
+			// if the value is empty, we may not have gotten the output of the
+			// container logs yet, and we should try again.
+			if val == "" {
+				return poll.Continue("output of container logs is empty")
```

Comment:
Let me think about this;

Instead of using the container's / service's logs to check if the value was set, could we abuse the container's command for this (something like `if cat /proc/sys/net/ipv4/ip_nonlocal_bind != expected value; exit 1`)?

Thinking a bit further about this; I'm wondering if we need to check this _at all_. Creating a container with custom `sysctl` settings is an existing feature. The only thing being added in this PR is that that container is now created through SwarmKit instead of `docker run`.

Because of that, I think all we need to verify is if the container-spec is correct (the expected sysctl options are set).

Reply:
> Instead of using the container's / service's logs to check if the value was set, could we abuse the container's command for this (something like if cat /proc/sys/net/ipv4/ip_nonlocal_bind != expected value; exit 1)?

What do you think is better about doing it this way? Having containers exit in a swarm service test leaves us dealing with swarmkit rescheduling those containers.

> Thinking a bit further about this; I'm wondering if we need to check this at all. Creating a container with custom `sysctl` settings is an existing feature. The only thing being added in this PR is that that container is now created through SwarmKit instead of `docker ru`n.

We do, actually, to make sure that the value of the Sysctls is plumbed through correctly. When writing this test, I actually missed a location where the value was plumbed through and so the value was making its way into the container spec, but not into the actual container.

Reply:
> We do, actually, to make sure that the value of the Sysctls is plumbed through correctly. 

Discussed this with @dperny on Slack; inspecting the container should contain enough information that the correct container-spec was created. Creating containers from a spec/config is already covered by other tests, so for this feature we don't have to test that part 👍